tests/pplat.cu

```diff
@@ -0,0 +1,256 @@
+        ASSERTDRV(cuDeviceGetAttribute(&dev_pci_device_id, CU_DEVICE_ATTRIBUTE_PCI_DEVICE_ID, dev));
+
+        cout  << "GPU id:" << n << "; name: " << dev_name 
+              << "; Bus id: "
+              << std::hex 
+              << std::setfill('0') << std::setw(4) << dev_pci_domain_id
+              << ":" << std::setfill('0') << std::setw(2) << dev_pci_bus_id
+              << ":" << std::setfill('0') << std::setw(2) << dev_pci_device_id
+              << std::dec
+              << endl;
+    }
+    cout << "selecting device " << dev_id << endl;
+    ASSERTDRV(cuDeviceGet(&dev, dev_id));
+
+    CUcontext dev_ctx;
+    ASSERTDRV(cuDevicePrimaryCtxRetain(&dev_ctx, dev));
+    ASSERTDRV(cuCtxSetCurrent(dev_ctx));
+
+    ASSERT_EQ(check_gdr_support(dev), true);
+
+    ASSERTDRV(galloc_fn(&mhandle, sizeof(*d_buf), true, true));
+    d_buf_cuptr = mhandle.ptr;
+    cout << "device ptr: 0x" << hex << d_buf_cuptr << dec << endl;
+
+    if (galloc_fn == gpu_mem_alloc)
+        cout << "gpu alloc fn: cuMemAlloc" << endl;
+    else
+        cout << "gpu alloc fn: cuMemCreate" << endl;
+
+    ASSERTDRV(cuMemsetD8(d_buf_cuptr, 0, sizeof(*d_buf)));
+
+    ASSERTDRV(cuMemHostAlloc((void **)&h_buf, sizeof(*h_buf), CU_MEMHOSTALLOC_PORTABLE | CU_MEMHOSTALLOC_DEVICEMAP));
+    ASSERT_NEQ(h_buf, (void*)0);
+    ASSERTDRV(cuMemHostGetDevicePointer(&h_buf_cuptr, h_buf, 0));
+    memset(h_buf, 0, sizeof(*h_buf));
+
+
+    gdr_t g = gdr_open_safe();
+
+    gdr_mh_t mh;
+    BEGIN_CHECK {
+        // tokens are optional in CUDA 6.0
+        ASSERT_EQ(gdr_pin_buffer(g, d_buf_cuptr, sizeof(*d_buf), 0, 0, &mh), 0);
+        ASSERT_NEQ(mh, null_mh);
+
+        void *map_d_ptr  = NULL;
+        ASSERT_EQ(gdr_map(g, mh, &map_d_ptr, sizeof(*d_buf)), 0);
+        cout << "map_d_ptr: " << map_d_ptr << endl;
+
+        gdr_info_t info;
+        ASSERT_EQ(gdr_get_info(g, mh, &info), 0);
+        cout << "info.va: " << hex << info.va << dec << endl;
+        cout << "info.mapped_size: " << info.mapped_size << endl;
+        cout << "info.page_size: " << info.page_size << endl;
+        cout << "info.mapped: " << info.mapped << endl;
+        cout << "info.wc_mapping: " << info.wc_mapping << endl;
+
+        // remember that mappings start on a 64KB boundary, so let's
+        // calculate the offset from the head of the mapping to the
+        // beginning of the buffer
+        int off = info.va - d_buf_cuptr;
+        cout << "page offset: " << off << endl;
+
+        d_buf = (uint32_t *)((uintptr_t)map_d_ptr + off);
+        cout << "user-space pointer: " << d_buf << endl;
+
+        cout << "CPU does gdr_copy_to_mapping and GPU writes back via cuMemHostAlloc'd buffer." << endl;
+        cout << "Running " << num_iters << " iterations with data size " << sizeof(*d_buf) << " bytes." << endl;
+
+        pp_kernel<<< 1, 1 >>>((uint32_t *)d_buf_cuptr, (uint32_t *)h_buf_cuptr, num_iters);
+
+        uint32_t i = 1;
+        while (READ_ONCE(*h_buf) != i) ;
+        MB();
```

Comment:
MB on Arm64 seems to be just a compiler barrier. Don't we need a stronger form?

Reply:
Nice catch! Yes, I updated the memory barrier instruction for ARM.